### DISCUSSION TITLE
Is it possible to programatically build ActivityProperty-ies?

### DISCUSSION THREAD

**Discussion by tomy2105** [elsa-workflows/elsa-core#828] [Q&A]
The ones I've seen built in by default hardcode ActivityProperty (used to display in UI) in code.
Is it possible to programatically build (or change) ActivityProperty of and activity fields?
For example during service initialization.

**Comment by VahidNaderi** [comment]
I'm not sure if I get your issue well, but isn't it the obvious way to inherit an activity in order to make changes to it?

**Comment by tomy2105** [comment]
Thanks for the response @VahidNaderi, however my question was a little deeper :).
After you make your activity class, inheriting from `Activity` your class will have public member properties.
Some of these are displayed in designer UI for that activity.
C# attribute called `ActivityProperty` is used to control how UI will render this property in the designer.
Example taken from HTTP activity.

`

        [ActivityProperty(

            UIHint = ActivityPropertyUIHints.Dropdown,

            Hint = "The content type to send with the request.",

            Options = new[] { "text/plain", "text/html", "application/json", "application/xml", "application/x-www-form-urlencoded" }

        )]

        public string? ContentType { get; set; }

`

My question is: is there a way to build this attribute and its content programatically in the runtime, for example during program/service initialization/startup and not have it hardcoded in the code?

**Comment by sfmskywalker** [comment]
That's most definitely possible. In fact, the actual set of activity properties is determined at runtime by things called *activity type providers*.

### Activity Type Provider

Elsa relies on `IActivityTypeProvider` implementations to collect the various types of activities available to the system by providing collections of `ActivityType` objects.

One such provider is the `TypeBasedActivityProvider`, which uses the activity types (types that implement `IActivity`) as registered with `ElsaOptions`.

### ActivityType

These are the heart and soul of an actual activity - they contain pointers to functions that actually run & resume activity logic.
The `TypeBasedActivityProvider` simply forwards the `ExecuteAsync` method on an activity class to the `ExecuteAsync` func property on an `ActivityType` instance.

In addition, an `ActivityType` has a func property called `Describe`, which returns an `ActivityDescriptor`.

### ActivityDescriptor

The `ActivityDescriptor` describes what **properties** an activity has. These properties are modeled as `ActivityPropertyDescriptor`. The `TypeBasedActivityProvider` relies on a service called `IDescribesActivityType` to reflect on the properties (annotated with `ActivityPropertyAttribute`) on an `IActivity` implementation and project them into `ActivityPropertyDescriptor` objects (which are simple POCOs).

### ActivityPropertyDescriptor

The `ActivityPropertyDescriptor` describes a single property of a given activity.

As you see, these are collected dynamically, which means _you can provide them in any way you like_.
Right now, if you wanted to customize what properties are displayed in the UI for a given activity type, perhaps a custom activity that you wrote, you would have to do a couple of things:

1. Annotate your activity class with `[Browasable(false)]` so that it does not get collected by the `TypeBasedActivityProvider` (at least, this is the desired behavior, but I believe this is broken at the time of this writing).
2. Create a custom `IActivityTypeProvider` implementation that provides an `ActivityType` with the desired behavior and type descriptor (which includes the properties you want).

When used in the designer, the user will see your custom activity properties, be able to provide values for them, and they will be stored as part of the workflow definition and provided at runtime when the workflow executed.

We might add more extensibility points to e.g. customize properties on pre-built activities by e.g. publishing mediator notifications, allowing your code to intercept & customize various aspects on the activity types and their properties. If you're interested in that sort of thing, then I recommend adding an issue with your use case so that we might get a better understanding of what we would need to do.

**Reply by tomy2105** [reply]
> We might add more extensibility points to e.g. customize properties on pre-built activities by e.g. publishing mediator notifications, allowing your code to intercept & customize various aspects on the activity types and their properties. If you're interested in that sort of thing, then I recommend adding an issue with your use case so that we might get a better understanding of what we would need to do.

I have voiced my specific need in https://github.com/elsa-workflows/elsa-core/issues/831. It is not for built-in ones (although that wouldn't be a bad idea either) but for custom ones.
